concent_api/core/payments/bankster.py

```diff
@@ -383,24 +390,82 @@ def sum_subtask_price(subtask_results_accepted_list: List[SubtaskResultsAccepted
 
 
 def get_provider_payment_info(
-    list_of_forced_payments: List[ForcedPaymentEvent],
-    list_of_payments: List[BatchTransferEvent],
+    list_of_settlement_payments: List[ForcedPaymentEvent],
+    list_of_transactions: List[BatchTransferEvent],
+    settlement_payment_claims: QuerySet,
     subtask_results_accepted_list: List[SubtaskResultsAccepted],
 ) -> Tuple[int, int]:
-    assert isinstance(list_of_payments, list)
-    assert isinstance(list_of_forced_payments, list)
+    assert isinstance(settlement_payment_claims, QuerySet)
+    assert isinstance(list_of_settlement_payments, list)
+    assert isinstance(list_of_transactions, list)
     assert isinstance(subtask_results_accepted_list, list)
 
-    force_payments_price = sum_payments(list_of_forced_payments)
-    payments_price = sum_payments(list_of_payments)
+    settlement_payment_total_price = sum_payments(list_of_settlement_payments)
+    payments_price = sum_payments(list_of_transactions)
+    satisfied_payments_price = settlement_payment_claims.aggregate(
+        sum_of_already_satisfied_claims=Coalesce(Sum('amount'), 0)
+    )['sum_of_already_satisfied_claims']
     subtasks_price = sum_subtask_price(subtask_results_accepted_list)
 
-    amount_paid = payments_price + force_payments_price
+    amount_paid = payments_price + satisfied_payments_price + settlement_payment_total_price
     amount_pending = subtasks_price - amount_paid
 
     return (amount_paid, amount_pending)
 
 
+def remove_duplicated_transactions_from_deposit_claims(
+    already_satisfied_claims: QuerySet,
+    list_of_settlement_payments: List[ForcedPaymentEvent],
+) -> QuerySet:
+    """
+    Removes duplicates from claims of settlement payments with pending transaction
+    which are on list of forced payments by adding exclusions to DepositClaim queryset.
+    """
+    settlement_payment_tx_hashes = [force_payment.tx_hash for force_payment in list_of_settlement_payments]
+
+    for matching_claim in already_satisfied_claims.filter(tx_hash__in=settlement_payment_tx_hashes):
+        assert matching_claim.concent_use_case == ConcentUseCase.FORCED_PAYMENT
+
+        matching_force_payment = [
+            force_payment for force_payment in list_of_settlement_payments
+            if force_payment.tx_hash == matching_claim.tx_hash
+        ][0]
```

Comment:
What if there's no claim corresponding to the payment in our database? We remove claims when their transactions are confirmed so it's a very common situation.

Reply:
Then we won't reach this place in code, because in this part we iterate only over claims that have `tx_hash` matched with one of the settlement payments from `list_of_settlement_payments`. So if there is `matching_claim`, it means that there must be object with the same `tx_hash` in `list_of_settlement_payments`.

Reply:
OK. You're right. It seems fine after all.

### Example
Just for reference, an example I created to analyze this case:

required confirmations: 6

payments:
- A (100 confirmations)
- B (80 confirmations)
- C (30 confirmations)
- D (6 confirmations)
- E (6 confirmations)
- F (5 confirmations)
- G (0 confirmations)
- H (transaction not submitted to the blockchain client yet)

claims in the database:
- 1 (hash=E)
- 2 (hash=F)
- 3 (hash=G)
- 4 (hash=None)

Then:
- `list_of_settlement_payments`: A, B, C, D, E
- `settlement_payment_tx_hashes`: A, B, C, D, E
- `already_satisfied_claims`: 1, 2, 3
- `already_satisfied_claims.filter(tx_hash__in=[A, B, C, D, E])`: 1

The case I was worried about was claim F or G (has claim with a hash but is not visible on the list obtained from the blockchain). But the filter takes care of it.